charms/slurmctld/src/interface_slurmd.py

```diff
@@ -286,3 +286,16 @@ def get_all_gres_info(self) -> Dict[str, Any]:
                             gres_info[node_name] = gres
 
         return gres_info
+
+    def get_active_nodes(self) -> List:
```

Comment:
[nit]: If we're grabbing just the name of the node, we could more narrowly scope this annotation to say that this method returns a list of strings rather than just a generic list:

```suggestion
    def get_active_nodes(self) -> List[str]:
```

Reply:
We need to account for the case where the empty list is returned.

Reply:
I see what you're saying, but per [PEP 484](https://peps.python.org/pep-0484/#the-typing-module), all containers with a uniform element type can be empty. This means that `list[str]` is the correct annotation, even if `get_active_nodes(...)` can possibly return an empty list. 

`pyright` will happily let through an empty list even if the annotation is `list[str]`. If we were intentionally return an empty list every time, then we would want to use the annotation `list[Never]`:

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByApgG5EgBQVAJkcFGkTAPoCGAxjEmcymLQGcAFAEooAWgB8UADZIBMANoKQAXQBcFKNqggmAVxAooi1VQq16rGQHdWcAcyLZ4zOQtETp7pcTJrNHV0DIxNVIA)

```python
from typing import Never


def get_active_nodes() -> list[str]:
    return []


def always_empty_list() -> list[Never]:
    return []
```

If we wanted to differentiate between a nullable and non-nullable case, we could use the annotation `list[str] | None`:

Code sample in [pyright playground](https://pyright-play.net/?code=CYUwZgBA5iAuD6BDAxrAlgNxPAdge1AGcAKQkAGzAEoIBaAPgnLUNgG1WAnAXQgB8IAOTw4QALgBQEaRHxEIAXghtuUmZzgBXTjlkEQhCGkjkQOYnIM1GABggUyQkSCA)

```python
def get_active_nodes(self) -> list[str] | None:
    nodes = []
    return nodes if len(nodes) > 0 else None
```


A bit wordy, but the TL;DR here is that `list[str]` is the correct annotation even if an empty list can be returned, and we should differentiate between `list[str]` and `None` if returning an empty list will cause the charm to break. I think we should use `list[str]` since it limits the types that can be added to the returned list in `get_active_nodes(...)`, and we get autocompletion benefits in our IDEs.